drivers/console/bm_serial.c

```diff
@@ -61,105 +57,293 @@ K_SEM_DEFINE(decode_sem, 0, 1);
-		if( byte_in == 0 )
-		{
-			// Frame complete
-			//LOG_INF( "Len=%d", len );
+		/* We reach this case if we receive non-preamble bytes when we expect them */
+		LOG_ERR("Expected Preamble, but got errant bytes");
+
+		/* Reset counter */
+		preamble_ctr = 0;
+	}
+    return ret;
+}
+
+bm_ret_t bm_serial_process_byte(uint8_t byte)
+{
+	bm_ret_t test_ret = {.retval=-EINPROGRESS, .length=0, .buf_ptr=NULL};
+	bm_parse_ret_t ret;
+	static uint16_t payload_ctr = 0;
+	static bm_parse_state_t _state = BM_ALIGN;
+	static uint8_t decoded_version = 0;
+	static uint8_t decoded_type = 0;
+	static uint16_t decoded_length = 0;
+	static uint16_t decoded_crc = 0;
+	int8_t decoded_nibble_lsb = 0;
+	int8_t decoded_nibble_msb = 0;
+	uint16_t computed_crc16 = 0;
+	bm_frame_header_t rx_header;
+
+	switch (_state)
+	{
+		case BM_ALIGN:
+		
+			/* Reset static variables */
+			decoded_length = 0;
+			decoded_version = 0;
+			decoded_type = 0;
+			decoded_crc = 0;
+			payload_ctr = 0;
+
+			ret = bm_serial_align(&byte);
+			_state = ret.new_state;
+			break;
+		case BM_COLLECT_HEADER:
+			encoded_rx_buf[write_buf_idx].buf[encoded_rx_ctr++] = byte;
 
-			if (len != 0)
+			/* Collect until we have a full frame header */
+			if (encoded_rx_ctr == (2*sizeof(bm_frame_header_t)))
 			{
-				// Decode
-				cobs_decode_result_t ret = cobs_decode( cobs_decoding_buf[write_buf_idx].buf, MAX_BM_FRAME_SIZE, input_buffer, len );
-				if( ret.status == COBS_DECODE_OK)
+				/* Start with version */
+				decoded_nibble_lsb = MAN_DECODE_TABLE[encoded_rx_buf[write_buf_idx].buf[0]];
+				decoded_nibble_msb = MAN_DECODE_TABLE[encoded_rx_buf[write_buf_idx].buf[1]];
+				if (decoded_nibble_lsb >= 0 || decoded_nibble_msb >= 0 )
+				{
+					decoded_version |= (((uint8_t) decoded_nibble_lsb) & 0xF);
+					decoded_version |= ((((uint8_t) decoded_nibble_msb) & 0xF) << 4);
+					rx_header.version = decoded_version;
+				}
+				else
 				{
-					//LOG_INF( "Decoded seq: %d", cobs_decoding_buf[write_buf_idx].buf[ sizeof(bm_frame_header_t)] );
-
-					cobs_decoding_buf[write_buf_idx].length = ret.out_len;
-
-					/* First check if the RX_Task has finished processing read_buf */
-					if ( k_sem_take(&processing_sem, K_NO_WAIT))
-					{
-						LOG_ERR("RX Task has not finished processing data. Overwriting current index of A/B Buffer.");
-					}
-					else
-					{
-						/* Flip idx of read and write buffers */
-						read_buf_idx = write_buf_idx;
-						write_buf_idx = 1 - write_buf_idx;
-					}
+					LOG_ERR("Invalid Manchester value for BM version");
+					/* Reset RX buffer index */
+					encoded_rx_ctr = 0;
+					/* Reset bm parse state */
+					_state = BM_ALIGN;
+					break;
+				}
 
-					/* Notify RX Task that frame is ready to be processed */
-					k_sem_give(&decode_sem);
+				/* Next decoded BM Payload Type */
+				decoded_nibble_lsb = MAN_DECODE_TABLE[encoded_rx_buf[write_buf_idx].buf[2]];
+				decoded_nibble_msb = MAN_DECODE_TABLE[encoded_rx_buf[write_buf_idx].buf[3]];
```

Comment:
I'd caution against using these hardcoded buffer indices. If you ever decide to change the bm_frame_header_struct, you'll have to go through and change each one of these carefully.

You can use the `offsetof` macro/function (yay C99!) to do this for you.

For example:

Instead of:
```
encoded_rx_buf[write_buf_idx].buf[0]
encoded_rx_buf[write_buf_idx].buf[1]
encoded_rx_buf[write_buf_idx].buf[2]
encoded_rx_buf[write_buf_idx].buf[3]
```
Do:

```
encoded_rx_buf[write_buf_idx].buf[offsetof(bm_frame_header_t, version) * 2]
encoded_rx_buf[write_buf_idx].buf[offsetof(bm_frame_header_t, version) * 2 + 1]
encoded_rx_buf[write_buf_idx].buf[offsetof(bm_frame_header_t, payload_type) * 2]
encoded_rx_buf[write_buf_idx].buf[offsetof(bm_frame_header_t, payload_type) * 2 + 1]
```

etc...

It's not as pretty, but it will be more robust to header changes.

Reply:
Oh wow, thanks for teaching me this. Never new this existed